scripts/jenkins/cloud/ansible/roles/heat-generator/library/generate_heat_model.py

```diff
@@ -374,6 +376,8 @@ def map_foreign_keys(root_element, element_name,
                     element_key)
 
     for sub_element_name, sub_element_schema in schema_sub_elements.items():
+        if element.get(sub_element_name, None) is None:
```

Comment:
I wrote this logic to re-organize the input model data structure to replace foreign keys with actual list or dictionaries containing references to other objects in the input model data structure. It started out as a simple thing, but turned out really convoluted, which I'm not really proud of.

I vaguely remember that the rest of the code might actually run on the assumption that there must be a  reference attribute for every foreign key relationship coded in the schema, even when the attribute is missing in the input model (e.g. every `networks` element must have a `network-group` attribute, even if the input model doesn't specify one, in which case its value will be set to None by this restructuring code). This change will break that assumption, but as long as you don't run into any other issues down the line, it should be fine. However, if I were you, I would still test the change against a bunch of generated input model using different scenarios and parameters, just to be sure :)

Reply:
I have tested it against a few models and it does the right thing...

Note that we only see a problem when an input model list object ends up with a None value, i.e. is given no value. This only happens when we, for example, disable all of the services that would run on that node type, so we don't generate a node type entry (e.g. for compute nodes) in the resources list...

This set of changes makes the heat model generation not fail in such cases... However the other changes in this patch set to fix the control_plane.yml and network_groups.yml rendering avoid generating models with such dangling entries.... So technically with those other changes in place we shouldn't hit these exceptions anymore...

Reply:
Are there specific model scenarios you have in mind for testing against?

Also note that this change won't explicitly set anything to None; it skips things if their value is none, leaving them untouched, rather than trying to dereference them on the assumption that they are a list.

Would you prefer that we just threw a more meaningful exception in such cases?

Note that if the generated model is not valid, it will eventually be caught by the config processor, if not the heat stack instantiation.

Reply:
I can't really think of any one particular use case where this change would be problematic, but I'm confident that the testing you've done would have revealed most weaknesses.